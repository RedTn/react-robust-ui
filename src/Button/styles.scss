$blue: blue;
$red: red;
$black: black;
$white: white;
$silver: silver;

.button{
    
}