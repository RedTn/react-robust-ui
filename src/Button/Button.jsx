import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scss';
import classNames from 'classnames/bind';
import Color from 'color';
import { commonPropTypes, commonDefaultProps } from '__internal/Utils/CommonProps.js';

const cx = classNames.bind(styles);

const applyDynamicStyles = ({ color = '', gradient = false, bordered = false }) => {
    const style = {};

    let baseColor = '';

    switch (color) {
        case 'blue':
            baseColor = styles['blue'];
            break;
        case 'red':
            baseColor = styles['red'];
            break;
        default:
            baseColor = styles.silver;
    }

    if (gradient) {
        style.background = `linear-gradient(${baseColor}, ${Color(baseColor).darken(0.1)})`;
    } else {
        style.background = baseColor;
    }

    style.color = Color(baseColor).isDark() ? styles.white : styles.black;

    if (bordered) {
        style.border = `1px solid ${Color(baseColor).darken(0.1)}`;
    }

    return style;
};

class Button extends Component {
  static propTypes = {
      ...commonPropTypes,
      children: PropTypes.node,
      onClick: PropTypes.func,
      color: PropTypes.string,
      gradient: PropTypes.bool,
      bordered: PropTypes.bool
  };

  static defaultProps = {
      ...commonDefaultProps,
      children: null,
      onClick: () => {},
      color: null,
      gradient: false,
      bordered: false
  };

  render() {
      const { className, children, color, gradient, bordered, style } = this.props;

      return (
          <button
              className={cx(className, 'button')}
              style={{ ...style, ...applyDynamicStyles({ color, gradient, bordered }) }}
          >
              {children}
          </button>
      );
  }
}

export default Button;
