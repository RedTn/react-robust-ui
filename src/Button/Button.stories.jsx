import React from 'react';
import { storiesOf } from '@storybook/react';
import Button from './Button.jsx';
import { action } from '@storybook/addon-actions';
import centered from '@storybook/addon-centered/react';
import { withKnobs, text, boolean, select } from '@storybook/addon-knobs';

storiesOf('Button', module)
    .addDecorator(withKnobs)
    .addDecorator(centered)
    .add('basic usage', () => (
        <Button
            color={select('Color', ['base', 'blue'], 'base')}
            gradient={boolean('Gradient', false)}
            bordered={boolean('Bordered', false)}
            onClick={action('clicked')}
        >
            {text('Text', 'Hello Button')}
        </Button>
    ));
