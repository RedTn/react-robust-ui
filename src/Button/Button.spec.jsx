import React from 'react';
import Button from './Button';
import { shallow } from 'enzyme';

describe('Button', () => {
    describe('integration', () => {
        it('renders', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button />);
            expect(wrapper).toMatchSnapshot();
        });
        it('does not apply bordered style when false', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="blue" bordered={false} />);
            const styles = wrapper.prop('style');
            expect(styles).toEqual(
                expect.not.objectContaining({
                    border: expect.anything()
                })
            );
        });
        it('does not bordered style when true', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="blue" bordered />);
            const styles = wrapper.prop('style');
            expect(styles).toEqual(
                expect.objectContaining({
                    border: expect.stringContaining('1px solid')
                })
            );
        });
        it('does not apply gradient style when false', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="blue" gradient={false} />);
            const styles = wrapper.prop('style');
            expect(styles).toEqual(
                expect.objectContaining({
                    background: expect.not.stringContaining('linear-gradient')
                })
            );
        });
        it('does apply gradient style when true', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="blue" gradient />);
            const styles = wrapper.prop('style');
            expect(styles).toEqual(
                expect.objectContaining({
                    background: expect.stringContaining('linear-gradient')
                })
            );
        });
        it('applies color style from prop', () => {
            expect.assertions(3);

            const wrapper = shallow(<Button gradient={false} />);
            expect(wrapper.prop('style')).toEqual(
                expect.objectContaining({
                    background: 'silver'
                })
            );
            wrapper.setProps({ color: 'blue' });
            expect(wrapper.prop('style')).toEqual(
                expect.objectContaining({
                    background: 'blue'
                })
            );
            wrapper.setProps({ color: 'red' });
            expect(wrapper.prop('style')).toEqual(
                expect.objectContaining({
                    background: 'red'
                })
            );
        });
        it('applies bright text color from a dark background', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="blue" gradient={false} />);
            expect(wrapper.prop('style')).toEqual(
                expect.objectContaining({
                    color: 'white'
                })
            );
        });
        it('applies dark text color from a bright background', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button color="base" gradient={false} />);
            expect(wrapper.prop('style')).toEqual(
                expect.objectContaining({
                    color: 'black'
                })
            );
        });
        it('applies default params to dynamic styling', () => {
            expect.assertions(1);

            const wrapper = shallow(<Button />);
            expect(wrapper.instance().applyDynamicStyles({})).toEqual(
                expect.objectContaining({
                    background: expect.anything(),
                    color: expect.anything()
                })
            );
        });
    });
    describe('unit', () => {});
});
