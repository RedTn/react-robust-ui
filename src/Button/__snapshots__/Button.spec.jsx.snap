// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button integration renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    bordered={false}
    className={null}
    color={null}
    gradient={false}
    onClick={null}
    style={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Typography
        className={null}
        inline={false}
        style={null}
        variant="button"
      />,
      "className": "button",
      "style": Object {
        "background": "silver",
        "color": "black",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": null,
        "className": null,
        "inline": false,
        "style": null,
        "variant": "button",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Typography
          className={null}
          inline={false}
          style={null}
          variant="button"
        />,
        "className": "button",
        "style": Object {
          "background": "silver",
          "color": "black",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": null,
          "className": null,
          "inline": false,
          "style": null,
          "variant": "button",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
